python/dgl/dataloading/dataloader.py

```diff
@@ -703,6 +703,8 @@ def __init__(self, graph, indices, graph_sampler, device='cpu', use_ddp=False,
                            for k, v in indices.items()}
                 indices_device = next(iter(indices.values())).device
             else:
+                if hasattr(indices, "device"):
+                    indices_device = indices.device
```

Comment:
Good catch.  But why does this solve the problem where `indices_device` is missing?  If the `indices` object does not have a `device` argument, then the variable `indices_device` will still not appear.  This will still cause problems in sanity checks below.

I guess a better approach will be either (1) setting `indices_device` to None in `except` block, then skip the checks involving `indices_device` if it is None (might be tricky if someone uses custom dataset and found the device not matching the device argument and graph's device), or (2) enforce the indices objects to have a `device` argument in case conversion to a torch tensor fails:

```python
try:
    reraise = False
    if ...:
        ...
    else:
        indices = torch.tensor(indices) if not torch.is_tensor(indices) else indices
        if not hasattr(indices, 'device'):
            reraise = True
            raise AttributeError('Custom indices dataset requires a 'device' attribute indicating where the indices is.')
        indices_device = indices.device
except:
    if reraise:
        raise
```

Reply:
I think in your solution, `torch.tensor(indices)` will failed when indices is a customize dataset, and `indices_device = indices.device` will not be executed. Since we don't know whether indices is a non tensor array-like nids or a customize dataset. Another solution is add a addition try before `indices = torch.tensor(indices)`. What about this solution?

Reply:
Looks good.